extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/CamelRecorder.java

```diff
@@ -69,23 +69,42 @@ public void addTypeConverterLoader(RuntimeValue<TypeConverterRegistry> registry,
             String version,
             CamelConfig config) {
 
-        FastCamelContext context = new FastCamelContext(
-                factoryFinderResolver.getValue(),
-                version,
-                xmlLoader.getValue(),
-                xmlModelDumper.getValue());
-
-        context.setDefaultExtension(RuntimeCamelCatalog.class, () -> new CamelRuntimeCatalog(config.runtimeCatalog));
-        context.setRegistry(registry.getValue());
-        context.setTypeConverterRegistry(typeConverterRegistry.getValue());
-        context.setLoadTypeConverters(false);
-        context.setModelJAXBContextFactory(contextFactory.getValue());
-        context.build();
-
-        // register to the container
-        beanContainer.instance(CamelProducers.class).setContext(context);
-
-        return new RuntimeValue<>(context);
+        if (config.main.lightweight) {
+            FastLightweightCamelContext context = new FastLightweightCamelContext(
+                    factoryFinderResolver.getValue(),
+                    version,
+                    xmlLoader.getValue(),
+                    xmlModelDumper.getValue());
+            context.setRuntimeCamelCatalog(new CamelRuntimeCatalog(config.runtimeCatalog));
+            context.setRegistry(registry.getValue());
+            context.setTypeConverterRegistry(typeConverterRegistry.getValue());
+            context.setLoadTypeConverters(false);
+            context.setModelJAXBContextFactory(contextFactory.getValue());
+            context.build();
+
+            // register to the container
+            beanContainer.instance(CamelProducers.class).setContext(context);
+
+            return new RuntimeValue<>(context);
+        } else {
+            FastCamelContext context = new FastCamelContext(
+                    null,
+                    factoryFinderResolver.getValue(),
+                    version,
+                    xmlLoader.getValue(),
+                    xmlModelDumper.getValue());
+            context.setRuntimeCamelCatalog(new CamelRuntimeCatalog(config.runtimeCatalog));
+            context.setRegistry(registry.getValue());
+            context.setTypeConverterRegistry(typeConverterRegistry.getValue());
+            context.setLoadTypeConverters(false);
+            context.setModelJAXBContextFactory(contextFactory.getValue());
+            context.build();
+
+            // register to the container
+            beanContainer.instance(CamelProducers.class).setContext(context);
+
+            return new RuntimeValue<>(context);
+        }
```

Comment:
This would be easier to read an maintain:

```suggestion
        final ExtendedCamelContext context;
        if (config.main.lightweight) {
            context = new FastLightweightCamelContext(
                    factoryFinderResolver.getValue(),
                    version,
                    xmlLoader.getValue(),
                    xmlModelDumper.getValue());
        } else {
            context = new FastCamelContext(
                    null,
                    factoryFinderResolver.getValue(),
                    version,
                    xmlLoader.getValue(),
                    xmlModelDumper.getValue());
        }

        context.setRuntimeCamelCatalog(new CamelRuntimeCatalog(config.runtimeCatalog));
        context.setRegistry(registry.getValue());
        context.setTypeConverterRegistry(typeConverterRegistry.getValue());
        context.setLoadTypeConverters(false);
        context.setModelJAXBContextFactory(contextFactory.getValue());
        context.build();

        // register to the container
        beanContainer.instance(CamelProducers.class).setContext(context);

        return new RuntimeValue<>(context);
```